File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/library/items/media/files/properties/playstats/UpdatePlayStatsOnCompleteRegistration.java
package com.lasthopesoftware.bluewater.client.library.items.media.files.properties.playstats;

import android.content.BroadcastReceiver;
import android.content.IntentFilter;

import com.lasthopesoftware.bluewater.client.connection.IConnectionProvider;
import com.lasthopesoftware.bluewater.client.connection.receivers.IConnectionDependentReceiverRegistration;
import com.lasthopesoftware.bluewater.client.library.items.media.files.properties.FilePropertiesProvider;
import com.lasthopesoftware.bluewater.client.library.items.media.files.properties.FilePropertiesStorage;
import com.lasthopesoftware.bluewater.client.library.items.media.files.properties.playstats.factory.PlaystatsUpdateSelector;
import com.lasthopesoftware.bluewater.client.library.items.media.files.properties.repository.FilePropertyCache;
import com.lasthopesoftware.bluewater.client.playback.service.broadcasters.PlaylistEvents;
import com.lasthopesoftware.bluewater.client.servers.version.ProgramVersionProvider;

import java.util.Collection;
import java.util.Collections;

public class UpdatePlayStatsOnCompleteRegistration implements IConnectionDependentReceiverRegistration {

	private static final Collection<IntentFilter> intents = Collections.singleton(new IntentFilter(PlaylistEvents.onFileComplete));

	@Override
	public BroadcastReceiver registerWithConnectionProvider(IConnectionProvider connectionProvider) {
		final FilePropertyCache cache = FilePropertyCache.getInstance();

		return new UpdatePlayStatsOnPlaybackCompleteReceiver(
			new PlaystatsUpdateSelector(
				connectionProvider,
				new FilePropertiesProvider(connectionProvider, cache),
				new FilePropertiesStorage(connectionProvider, cache),
				new ProgramVersionProvider(connectionProvider)));
	}

	@Override
	public Collection<IntentFilter> forIntents() {
		return intents;
	}


}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/browsing/files/properties/playstats/UpdatePlayStatsOnCompleteRegistration.kt
package com.lasthopesoftware.bluewater.client.browsing.files.properties.playstats

import com.lasthopesoftware.bluewater.client.browsing.files.properties.ScopedFilePropertiesProvider
import com.lasthopesoftware.bluewater.client.browsing.files.properties.playstats.factory.PlaystatsUpdateSelector
import com.lasthopesoftware.bluewater.client.browsing.files.properties.repository.FilePropertyCache
import com.lasthopesoftware.bluewater.client.browsing.files.properties.storage.ScopedFilePropertiesStorage
import com.lasthopesoftware.bluewater.client.browsing.library.revisions.ScopedRevisionProvider
import com.lasthopesoftware.bluewater.client.connection.IConnectionProvider
import com.lasthopesoftware.bluewater.client.connection.authentication.ScopedConnectionAuthenticationChecker
import com.lasthopesoftware.bluewater.client.connection.receivers.RegisterReceiverForEvents
import com.lasthopesoftware.bluewater.client.playback.service.broadcasters.messages.PlaybackMessage.TrackCompleted
import com.lasthopesoftware.bluewater.client.servers.version.ProgramVersionProvider
import com.lasthopesoftware.bluewater.shared.cls
import com.lasthopesoftware.bluewater.shared.messages.application.ApplicationMessage
import com.lasthopesoftware.bluewater.shared.messages.application.SendApplicationMessages

class UpdatePlayStatsOnCompleteRegistration(private val messageBus: SendApplicationMessages) : RegisterReceiverForEvents {
	override fun registerWithConnectionProvider(connectionProvider: IConnectionProvider): (ApplicationMessage) -> Unit {
		val cache = FilePropertyCache.getInstance()
		val scopedRevisionProvider = ScopedRevisionProvider(connectionProvider)
		return UpdatePlayStatsOnPlaybackCompleteReceiver(
			PlaystatsUpdateSelector(
				connectionProvider,
				ScopedFilePropertiesProvider(connectionProvider, scopedRevisionProvider, cache),
				ScopedFilePropertiesStorage(
					connectionProvider,
					ScopedConnectionAuthenticationChecker(connectionProvider),
					scopedRevisionProvider,
					cache,
					messageBus,
				),
				ProgramVersionProvider(connectionProvider)
			)
		)
	}

	override fun forClasses(): Collection<Class<out ApplicationMessage>> = classes

	companion object {
		private val classes = setOf<Class<out ApplicationMessage>>(cls<TrackCompleted>())
	}
}
